proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+            path: kubexec.sh
+            mode: 365
+```
+
+### Resulting Launcher
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ${job-id}-launcher
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: kubectl-delivery
+        image: rongou/kubectl-delivery:latest
+        volumeMounts:
+        - name: ml-job-kubectl
+          mountPath: /opt/kube
+        env:
+        - name: TARGET_DIR
+          value: /opt/kube
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+        resources:
+          limits:
+            nvidia.com/gpu: 8
+        volumeMounts:
+        - name: ml-job-kubectl
+          mountPath: /opt/kube
+        - name: ml-job-config
+          mountPath: /etc/mpi
+        env:
+        - name: OMPI_MCA_plm_rsh_agent
+          value: /etc/mpi/kubexec.sh
+        - name: OMPI_MCA_orte_default_hostfile
+          value: /etc/mpi/hostfile
+      restartPolicy: OnFailure
+      serviceAccountName: ${job-id}-launcher
+      volumes:
+      - name: ml-job-kubectl
+        emptyDir: {}
+      - name: ml-job-config
+        configMap:
+          name: ${job-id}-config
+          items:
+          - key: kubexec.sh
+            path: kubexec.sh
+            mode: 365
+          - key: hostfile
+            path: hostfile
+            mode: 292
+```
+
+The launcher pod invokes `mpirun` and communicates with worker pods through MPI.
+The initial handshake is done through `kubectl exec` instead of SSH. Logs can
+be accessed through the launcher pod.
+
+## Design
+We create a new custom controller that listens for `MPIJob` resources. When a
+new `MPIJob` is created, the controller goes through the following *logical*
+steps:
+1.  Create a `ConfigMap` that contains:
+    *   A helper shell script that can be used by `mpirun` in place of ssh. It
+        invokes `kubectl exec` for remote execution.
+    *   A `hostfile` that lists the pods in the worker `StatefulSet` (in the
+        form of `${job-id}-worker-0`, `${job-id}-worker-1`, ...), and the
+        available slots (GPUs) in each pod.
+1.  Create the RBAC resources (`Role`, `ServiceAccount`, `RoleBinding`) to allow
+    remote execution (`pods/exec`).
+1.  Create the worker `StatefulSet` that contains the desired replicas minus 1,
+    as the launcher also does work. The command is set to sleep indefinitely.
+1.  Wait for the worker pods to be ready.
+1.  Create the launcher Job. It’s run under the `ServiceAccount` created in
+    step 2. Environment variables are set to tell `mpirun` to use our helper
+    script for remote execution, and the `hostfile` to use. The `kubectl` binary
+    is delivered to an `emptyDir` volume through an init container.
+1.  After the launcher job finishes, set the `replicas` to 0 in the worker
+    `StatefulSet`.
+
+![MPI Operator](diagrams/mpi-operator.png)
+
+It may be desirable to schedule all the GPUs in a single Kubernetes resource
```

Comment:
That sounds like a good default option.

Reply:
Do you all think this should be the only option? If no one ever wants to allocate GPUs to the launcher, then we can just get rid of that mode, to keep things simple.

Reply:
I think it could be an only option.